### ISSUE TITLE
Adding and then sorting an array breaks repeat binding

### ISSUE THREAD

**Issue by simonfox** [aurelia/binding#233]
When adding (`push`ing) to an array and then sorting the result, a repeat binding doesn't update correctly.

I have created a repro [here](https://github.com/simonfox/skeleton-navigation-sort-bug). I have added a `repeat` binding over a collection to the `Welcome` view model. There are a couple buttons below, one to add then sort and one to add without sort. If you use add with sort, you will see the unexpected behavior.

**Comment by devmondo** [comment]
i am facing the same issue

**Comment by BruceL33t** [comment]
Ye, this is a huge problem for me as well :/ I can't to this.somearray[0] = something, since it is not reflected in the views. On the other hand, pushing works just fine. this.somearray.push(something) works... I hope this will be fixed quickly

**Comment by jdanyow** [comment]
@BruceL33t your issue is different than what @simonfox reported.  Aurelia cannot observe indexed assignment of an array- at least not without Object.observe, which is no longer being championed in TC39.  As long as you use methods on the array instance to mutate the array it will work as you expect.  In your case, change `this.somearray[index] = something` to `this.somearray.splice(index, 1, something)`.

**Comment by vuorinem** [comment]
I was updating my project to the latest release and also faced this issue (made a simple plunkr repro: http://plnkr.co/edit/k3SrDHFKt9x0Wf0O5Ect). Is there an easy workaround for this, and is there a fix planned?

**Comment by martingust** [comment]
This is what I'm experiencing when doing push and sort on the underlying array:

The repeat gets two change records. One for adding one view (with index inside the bounds of the array). Another for removing one view, on max index +2. 

`view-slot` before making any changes
![screen shot 2015-11-23 at 21 03 00](https://cloud.githubusercontent.com/assets/5166897/11348743/95700eea-9228-11e5-9688-f0938d9a5582.png)

Second splice index off by +1?
![screen shot 2015-11-23 at 21 02 34](https://cloud.githubusercontent.com/assets/5166897/11348730/89412046-9228-11e5-88e3-214213022825.png)

@jdanyow / @EisenbergEffect Any ideas? Potentially looks like the splices are calculated wrong somewhere in [calcSplices](https://github.com/aurelia/binding/blob/bbcabc52b299df4a0b4335fe2ec5df9aa9cd38a9/src/array-change-records.js#L146).

**Comment by jdanyow** [comment]
I think @simonfox found a case where [this](https://github.com/aurelia/binding/blob/master/src/collection-observation.js#L83) is needed.

**Comment by martingust** [comment]
@jdanyow Concatenating  `changeRecords` and `records` did fix the error. But the sorting in the DOM is off. 

On [this](https://github.com/aurelia/binding/blob/master/src/collection-observation.js#L85) line I added this:

```javascript
if (changeRecords) {
  records = changeRecords.concat(records);
}
```

[Here](https://github.com/aurelia/templating-resources/commit/8769ce34d44746ccc65289a6c9037c4fc2d879b5) is a failing repeat integration test.

Let me know if you have any ideas. I can continue on this tomorrow.

**Comment by EisenbergEffect** [comment]
@jdanyow @martingust I wanted to push out a release today or tomorrow since we've built up a number of fixes. Any idea if this could be fixed by tomorrow? I just want to know if I should wait or not.

**Comment by martingust** [comment]
@EisenbergEffect I will continue look into this tomorrow, but I have no idea right now what the issue would be so I can't say if it will be fixed or not by then.

**Comment by EisenbergEffect** [comment]
No problem. The master is safe to release as is though, correct?

**Comment by jdanyow** [comment]
yep- all tests passing (other than this one)

**Comment by EisenbergEffect** [comment]
Ok, great. I'll get a release out and then we can get this one in as soon as it's done in a later patch.

**Comment by martingust** [comment]
@EisenbergEffect @jdanyow  I continued trying to trace this down. This is what I have found so far:

The bug happens when doing `push` and (certain) `sort` on arrays when the `repeat`'s [views require lifecycle](https://github.com/aurelia/templating-resources/blob/master/src/array-repeat-strategy.js#L99). As in earlier comment I tried to concat records with existing records [here](https://github.com/aurelia/binding/blob/master/src/collection-observation.js#L83) that fixes the error but the existing records gets duplecated in the `viewSlot`.

```javascript
 // initial array
 var foo = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

 // doing push and sort
 foo.push('10');
 foo.sort((a, b) => {});

console.log(foo); // ["0", "10", "2", "3", "4", "1", "6", "7", "8", "9", "5"]
```

The `repeat` gets 3 splices, where each splice have 1 add and 1 delete. First the views get deleted. The first two splices correctly deletes at index 1 ('1') and index 5 ('5') third wants to delete at index 10 (outside array), and error is thrown. 

If doing an ugly hack, to skip deleting if index is outside of array, the `repeat` processes and adds all views to be added, correctly.

In other words; the problem (as I see it) is with the third splice, which should not contain any deletes, just 1 add. I have looked through `array-change-records.js` without any luck figuring out if something goes wrong here or not. Wild guess is that something is going wrong in there, when calculating the edit distances, perhaps something to do with [this](https://github.com/aurelia/binding/blob/master/src/array-change-records.js#L27) comment. I'm not very familiar with this code and not making much progress from here, would need you guys to take a look.

**Comment by jdanyow** [comment]
@martingust I merged in the integration test you put together and added logic to the array observation to force a flush of the change records prior to doing an array operation that requires a "reset" (sort or reverse).  Downside is this will cause DOM changes outside of the task queue but I couldn't think of a way to get around it.

**Comment by martingust** [comment]
@jdanyow Awesome. I think we can live with that, `repeat` performance is pretty good in any case.

**Comment by simonfox** [comment]
@EisenbergEffect any chance of getting a patch release for this fix this week?

**Comment by EisenbergEffect** [comment]
Yes, I'm waiting to hear about a couple of other issues.

**Comment by simonfox** [comment]
Awesome thanks @EisenbergEffect

**Comment by ben-girardet** [comment]
I'm coming across a similar issue and can't get it to work. In an array (`this.steps`): I need to move an item from a position to another (index `previousPosition` to `newPosition`). This array is used in a repeater in the view.

Here are the relevant parts of the code I've tried:

**steps.js** (VM)
```
// Option 1 (using splice)
this.steps.splice(newPosition, 0, this.steps.splice(previousPosition, 1)[0]);

// Option 2 (using sort)
let index = 0;
for(let step of this.steps) {
  step.position = (index < newPosition) ? index : index+1;
  if(index == previousPosition) {
    step.position = newPosition;
  }
  index++;
}
this.steps.sort((a,b)=>{
  return a.position - b.position
});
```

**steps.html**
```
<li repeat.for="step of steps">...</li>
```

Both options have problems when the view gets updated. Sometime the moved item gets duplicated, sometimes the array gets an empty new item, sometimes an item gets missing, ...

Is this stil a bug in the binding system or am I doing something wrong when updating the array ?

**Comment by EisenbergEffect** [comment]
@jdanyow @martingust Can you guys look into this?

**Comment by Tarpsvo** [comment]
Experiencing the exact issue @martingust described in detail here earlier.

**Comment by ohjames** [comment]
I'm seeing this also. More generally I'm seeing Aurelian fail to maintain the array order with the order on screen if I do many splices and pushes in a row. :(

**Comment by ohjames** [comment]
Oh and latest version here too.

**Comment by ohjames** [comment]
My perspective on the importance of this bug is that Aurelia has never been able to maintain order of my lists in the view due to various bugs, two of which have been fixed but this one remains.

My website is all about sorting things by score. In the year I've been using Aurelia it's never been fit for this simplest of purposes. That coupled with dealing with so many bugs and a lack of documentation has left me feeling extremely regretful about my choice.
